### PR TITLE
Implement UI host unification tests

### PR DESCRIPTION
**Pull request by abalakh**
This PR:
* Implements all the host unification stubs we have
* Adds Content Host UI entity (as it's required for host unification tests)
* updates `VM.register_contenthost()` method to register content host not only by organization name + activation key, but also by credentials+organization+environment name

Tests results:
```python
py.test tests/foreman/ui/test_hostunification.py -n 2 -v
============================= test session starts ==============================
platform linux2 -- Python 2.7.10, pytest-2.9.2, py-1.4.31, pluggy-0.3.1 -- /home/andrii/env/bin/python
cachedir: .cache
rootdir: /home/andrii/workspace/robottelo, inifile: 
plugins: xdist-1.13.1, repeat-0.2
gw0 I / gw1 I
[gw0] linux2 Python 2.7.10 cwd: /home/andrii/workspace/robottelo
[gw1] linux2 Python 2.7.10 cwd: /home/andrii/workspace/robottelo
[gw0] Python 2.7.10 (default, Sep  8 2015, 17:20:17)  -- [GCC 5.1.1 20150618 (Red Hat 5.1.1-4)]
[gw1] Python 2.7.10 (default, Sep  8 2015, 17:20:17)  -- [GCC 5.1.1 20150618 (Red Hat 5.1.1-4)]
gw0 [10] / gw1 [10]

scheduling tests via LoadScheduling

tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_delete_from_allhosts <- robottelo/decorators.py 
tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_negative_add_contents_to_unregistered_host <- robottelo/decorators.py 
[gw0] PASSED tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_negative_add_contents_to_unregistered_host <- robottelo/decorators.py 
tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_negative_add_subs_to_unregistered_host <- robottelo/decorators.py 
[gw0] PASSED tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_negative_add_subs_to_unregistered_host <- robottelo/decorators.py 
tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_re_register_host <- robottelo/decorators.py 
[gw1] PASSED tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_delete_from_allhosts <- robottelo/decorators.py 
tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_provision_foreman_host <- robottelo/decorators.py 
[gw1] PASSED tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_provision_foreman_host <- robottelo/decorators.py 
tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_register_host_via_rhsm <- robottelo/decorators.py 
[gw1] PASSED tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_register_host_via_rhsm <- robottelo/decorators.py 
tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_rename_content_host <- robottelo/decorators.py 
[gw0] PASSED tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_re_register_host <- robottelo/decorators.py 
tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_register_host_via_ak <- robottelo/decorators.py 
[gw1] PASSED tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_rename_content_host <- robottelo/decorators.py 
[gw0] PASSED tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_register_host_via_ak <- robottelo/decorators.py 
tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_rename_foreman_host <- robottelo/decorators.py 
tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_unregister_content_host <- robottelo/decorators.py 
[gw1] PASSED tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_rename_foreman_host <- robottelo/decorators.py 
[gw0] PASSED tests/foreman/ui/test_hostunification.py::HostContentHostUnificationTestCase::test_positive_unregister_content_host <- robottelo/decorators.py 

========================= 10 passed in 747.54 seconds ==========================
```

### CODE REVIEW

File: robottelo/ui/contenthost.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        """Specify locator for Content Hosts entity search procedure"""
+        return locators['contenthost.select_name']
+
+    def add_subscriptions(self, subscriptions=None, tab_locator=None,
+                          select_locator=None):
+        """Add or remove subscription association for content host."""
+        strategy, value = locators['contenthost.subscription_select']
+        self.click(tab_locators['contenthost.tab_subscriptions'])
+        self.click(tab_locators['contenthost.tab_subscriptions_subscriptions'])
+        if not self.find_element(tab_locator):
```
Comment (oshtaier) [inline robottelo/ui/contenthost.py]: I suggest to use `self.wait_until_element` instead `find_element` to have more stable runs in future

File: robottelo/ui/contenthost.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        self.click(locators['contenthost.unregister'])
+        if really:
+            self.click(common_locators['confirm_remove'])
+        else:
+            self.click(common_locators['cancel'])
+
+    def validate_subscription_status(self, name, expected_value=True,
+                                     timeout=120):
+        """Check whether a content host has active subscription or not"""
+        for _ in range(timeout / 5):
```
Comment (oshtaier) [inline robottelo/ui/contenthost.py]: can we omit that division operation?

File: tests/foreman/ui/test_hostunification.py [hunk 3]
```diff
@@ -67,32 +93,33 @@ def test_positive_register_host_via_ak(self):
...
+            self.assertEqual(result.return_code, 0)
+            with Session(self.browser) as session:
+                session.nav.go_to_select_org(org.name)
+                self.assertIsNotNone(self.hosts.search(vm.hostname))
+                self.assertIsNotNone(self.contenthost.search(vm.hostname))
 
-    @stubbed()
-    @run_only_on('sat')
-    @tier3
-    def test_positive_validate_org(self):
```
Comment (oshtaier) [inline tests/foreman/ui/test_hostunification.py]: Please, explain the reason of removing that stub

File: tests/foreman/ui/test_hostunification.py [hunk 4]
```diff
@@ -129,12 +188,31 @@ def test_positive_rename_foreman_host(self):
...
+                session.nav.go_to_select_org(self.org_.name)
+                name, domain_name = vm.hostname.split('.', 1)
+                new_name = gen_string('alphanumeric').lower()
+                self.hosts.update(
+                    name,
+                    domain_name,
+                    new_name=new_name,
+                )
+                # Host rename operation is not atomic and may take some time
+                for _ in range(3):
```
Comment (oshtaier) [inline tests/foreman/ui/test_hostunification.py]: Not sure that I get when second iteration of that loop can happen
if searched not None -> leave loop
if None -> fail the test

File: tests/foreman/ui/test_hostunification.py [hunk 5]
```diff
@@ -252,7 +435,40 @@ def test_positive_add_contents_to_unregistered_host(self):
...
+                interface_parameters=[
+                    ['Type', 'Interface'],
+                    ['MAC address', host.mac],
+                    ['Domain', host.domain.name],
+                    ['Primary', True],
+                ],
+            )
+            hostname = u'{0}.{1}'.format(host.name, host.domain.name)
+            self.contenthost.install_package(hostname, 'busybox')
+            self.assertIsNotNone(common_locators['alert.error'])
```
Comment (oshtaier) [inline tests/foreman/ui/test_hostunification.py]: `self.assertIsNotNone(common_locators['alert.error'])` locator will never be None :) maybe you wanted to wait for element with that locator?

**Comment by oshtaier** [comment]
ACK pending questions

File: tests/foreman/ui/test_hostunification.py (hunk 3 above)
Comment (abalakh) [inline tests/foreman/ui/test_hostunification.py]: All the hosts we register via rhsm have organization specified, for example [here](https://github.com/SatelliteQE/robottelo/pull/3577/files#diff-25c1fd038f0ca18fe9cded0814312d79R74) or [here](https://github.com/SatelliteQE/robottelo/pull/3577/files#diff-25c1fd038f0ca18fe9cded0814312d79R116). So basically this test is already covered in any other host registration test. And it's not an option not to specify org in other tests - we won't be able to see the host on ui, as the user is forced to select organization to see hosts :)

File: tests/foreman/ui/test_hostunification.py (hunk 4 above)
Comment (abalakh) [inline tests/foreman/ui/test_hostunification.py]: Forgot to use `finally:` here. Good catch!

File: robottelo/ui/contenthost.py (hunk 2 above)
Comment (abalakh) [inline robottelo/ui/contenthost.py]: Like this?
```python
foo = 0
while foo < timeout:
    ...
    foo += 5
```
I don't mind it, but what's wrong with division, just curious?

Comment (oshtaier) [inline robottelo/ui/contenthost.py]: like `timeout=20` or I missed something

Comment (abalakh) [inline robottelo/ui/contenthost.py]: With `timeout=20` and without division it would be attempts counter, not timeout.

**Comment by abalakh** [comment]
Comments addressed. Thanks @oshtaier for few good catches!
```
py.test tests/foreman/ui/test_hostunification.py -k 'test_negative_add_contents_to_unregistered_host'
================================ test session starts =================================
platform linux2 -- Python 2.7.10, pytest-2.9.2, py-1.4.31, pluggy-0.3.1
rootdir: /home/andrii/workspace/robottelo, inifile: 
plugins: xdist-1.13.1, repeat-0.2
collected 10 items 

tests/foreman/ui/test_hostunification.py .

===== 9 tests deselected by '-ktest_negative_add_contents_to_unregistered_host' ======
====================== 1 passed, 9 deselected in 59.04 seconds =======================
```

File: robottelo/vm.py [hunk 6]
```diff
@@ -230,9 +230,13 @@ def install_katello_ca(self):
...
-    def register_contenthost(self, activation_key, org, force=True,
-                             releasever=None):
-        """Registers content host on foreman server using activation-key.
+    def register_contenthost(self, org, activation_key=None, lce=None,
+                             force=True, releasever=None):
+        """Registers content host on foreman server using activation-key. This
+        can be done in two ways: either by specifying organization name and
+        activation key name or by specifying organization name and lifecycle
+        environment name (administrator credentials for authentication will be
+        passed automatically)
```
Comment (sghai) [inline robottelo/vm.py]: :100: for covering both scenarios and with nice explanation.

**Comment by sghai** [comment]
Awesome stuff. These are much needed tests and covers lots of areas like hosts/content-hosts and client operations like (register/unregister). 

A few non-blocking comments for this PR:
- I see `unregister` method to unregister the content-host, but not the delete method to remove content-host. That's a new change in 6.2. When you click 'unregister' you will get two options a) to unsubscribe the hosts from subscriptions b) to completely remove the content-host. So let's add that test too in another PR.

- I see `install_package` method, would love to see the test-coverage on this..thanks :)

**Comment by oshtaier** [comment]
@sghai Thanks for suggestions! Agreed with Andrii that he look on that right after current PR is merged

**Comment by oshtaier** [comment]
ACK

**Pull request merged**